liquid-staking/src/funds/claim.rs

```diff
@@ -0,0 +1,80 @@
+multiversx_sc::imports!();
+
+use crate::{
+    basics::constants::{DEFAULT_GAS_TO_CLAIM_REWARDS, DEFAULT_MIN_GAS_TO_SAVE_PROGRESS},
+    basics::errors::{ERROR_NOT_ACTIVE, ERROR_NO_DELEGATION_CONTRACTS},
+    config::{self},
+    delegation::{self, ClaimStatusType},
+    delegation_proxy, StorageCache,
+};
+
+#[multiversx_sc::module]
+pub trait ClaimModule:
+    config::ConfigModule
+    + multiversx_sc_modules::default_issue_callbacks::DefaultIssueCallbacksModule
+    + delegation::DelegationModule
+{
+    #[endpoint(claimRewards)]
+    fn claim_rewards(&self) {
+        let storage_cache = StorageCache::new(self);
+
+        require!(
+            self.is_state_active(storage_cache.contract_state),
+            ERROR_NOT_ACTIVE
+        );
+
+        let delegation_addresses_mapper = self.delegation_addresses_list();
+        require!(
+            !delegation_addresses_mapper.is_empty(),
+            ERROR_NO_DELEGATION_CONTRACTS
+        );
+        let claim_status_mapper = self.delegation_claim_status();
+        let old_claim_status = claim_status_mapper.get();
+        let current_epoch = self.blockchain().get_block_epoch();
+
+        self.check_claim_operation(old_claim_status, current_epoch);
+        let mut delegation_addresses = self.addresses_to_claim();
+        if delegation_addresses.is_empty() {
+            self.prepare_claim_operation();
+        }
+
+        while !delegation_addresses.is_empty() {
+            let gas_left = self.blockchain().get_gas_left();
+            if gas_left < DEFAULT_MIN_GAS_TO_SAVE_PROGRESS {
+                break;
+            }
+
+            let current_node = delegation_addresses.pop_back().unwrap();
+            let address = current_node.clone().into_value();
+
+            self.tx()
+                .to(&address)
+                .typed(delegation_proxy::DelegationMockProxy)
+                .claim_rewards()
+                .gas(DEFAULT_GAS_TO_CLAIM_REWARDS)
+                .callback(ClaimModule::callbacks(self).claim_rewards_callback())
+                .register_promise();
```

Comment:
Can be its own function.

Reply:
gets too complicated because I need to return a `ManagedOption<LinkedListNode<ManagedAddress>>` where if it's  None we need to break while otherwise unpack and call callback..

I think it looks great like this :nutu:

Reply:
I missed that. The only lines 47-56.